core/src/main/scala/caliban/execution/Executor.scala

```diff
@@ -178,20 +179,34 @@ object Executor {
   }
 
   private[caliban] def mergeFields(field: Field, typeName: String): List[Field] = {
-    val allFields = field.fields ++ field.conditionalFields.getOrElse(typeName, Nil)
-    allFields
-      .foldLeft(ListMap.empty[String, Field]) {
-        case (result, field) =>
-          val name = field.alias.getOrElse(field.name)
-          result.updated(
-            name,
-            result
-              .get(name)
-              .fold(field)(f => f.copy(fields = f.fields ++ field.fields))
-          )
+    // ugly mutable code but it's worth it for the speed ;)
+    val array = ArrayBuffer.empty[Field]
+    val map   = collection.mutable.Map.empty[String, Int]
+    var index = 0
+
+    field.fields.foreach { field =>
+      if (field.condition.forall(_ == typeName)) {
+        val name = field.alias.getOrElse(field.name)
+        map.get(name) match {
+          case None =>
+            // first time we see this field, add it to the array
+            array += field
+            index += 1
+          case Some(index) =>
+            // field already existed, merge it
+            val f = array(index)
+            array(index) = f.copy(fields = f.fields ++ field.fields)
+        }
       }
-      .values
-      .toList
+    }
+
+    val result = List.newBuilder[Field]
+    var i      = 0
+    while (i < index) {
+      result += array(i)
+      i += 1
+    }
+    result.result()
```

Comment:
I propose some improvements over the current implementation:
```scala
  private[caliban] def mergeFields(field: Field, typeName: String): List[Field] = {
    // ugly mutable code but it's worth it for the speed ;)
    val array = ArrayBuffer.empty[Field]
    val map   = collection.mutable.Map.empty[String, Int]

    val reusedFieldBuilder: ArrayBuffer[Field] = ArrayBuffer.empty[Field]

    field.fields.foreach { field =>
      if (field.condition.forall(_ == typeName)) {
        val name = field.alias.getOrElse(field.name)
        map.get(name) match {
          case None =>
            // first time we see this field, add it to the array
            array += field
          case Some(index) =>
            // field already existed, merge it
            val f = array(index)
            reusedFieldBuilder ++= f.fields
            reusedFieldBuilder ++= field.fields
            array(index) = f.copy(fields = reusedFieldBuilder.toList)
            reusedFieldBuilder.clear()
        }
      }
    }

    array.toList
  }
```

Comments:
- `ArrayBuffer` does not require tracking its size manually (unlike preallocated arrays)
- `array(index) = f.copy(fields = f.fields ++ field.fields)` could be improved.
- Having a `ListBuilder` at the end is unnecessary, conversion to a `List` is already optimized (inside `IndexedSeqOptimized`)

Reply:
Thanks! Applied your suggestion.

Reply:
Hmm... come to think of it, maybe just ::: could be better than using a reusedFieldBuilder and make the code simpler. WDYT?

Reply:
Yeah maybe. It didn’t make a difference in benchmarks anyway.

Reply:
Simplified. This branch isn't even in the hot path.